commit.go

```diff
@@ -358,45 +356,50 @@ func (b *Builder) Commit(ctx context.Context, dest types.ImageReference, options
 	if len(options.AdditionalTags) > 0 {
 		switch dest.Transport().Name() {
 		case is.Transport.Name():
-			img, err := is.Transport.GetStoreImage(b.store, dest)
-			if err != nil {
-				return imgID, nil, "", fmt.Errorf("locating just-written image %q: %w", transports.ImageName(dest), err)
-			}
-			if err = util.AddImageNames(b.store, "", systemContext, img, options.AdditionalTags); err != nil {
-				return imgID, nil, "", fmt.Errorf("setting image names to %v: %w", append(img.Names, options.AdditionalTags...), err)
+			if dest.Transport().Name() == is.Transport.Name() {
+				_, img, err := is.ResolveReference(dest)
+				if err != nil {
+					return imgID, nil, "", fmt.Errorf("locating just-written image %q: %w", transports.ImageName(dest), err)
+				}
+				if err = util.AddImageNames(b.store, "", systemContext, img, options.AdditionalTags); err != nil {
+					return imgID, nil, "", fmt.Errorf("setting image names to %v: %w", append(img.Names, options.AdditionalTags...), err)
+				}
+				logrus.Debugf("assigned names %v to image %q", img.Names, img.ID)
 			}
-			logrus.Debugf("assigned names %v to image %q", img.Names, img.ID)
 		default:
 			logrus.Warnf("don't know how to add tags to images stored in %q transport", dest.Transport().Name())
 		}
 	}
 
-	img, err := is.Transport.GetStoreImage(b.store, dest)
-	if err != nil && !errors.Is(err, storage.ErrImageUnknown) {
-		return imgID, nil, "", fmt.Errorf("locating image %q in local storage: %w", transports.ImageName(dest), err)
-	}
-	if err == nil {
-		imgID = img.ID
-		toPruneNames := make([]string, 0, len(img.Names))
-		for _, name := range img.Names {
-			if nameToRemove != "" && strings.Contains(name, nameToRemove) {
-				toPruneNames = append(toPruneNames, name)
-			}
-		}
-		if len(toPruneNames) > 0 {
-			if err = b.store.RemoveNames(imgID, toPruneNames); err != nil {
-				return imgID, nil, "", fmt.Errorf("failed to remove temporary name from image %q: %w", imgID, err)
-			}
-			logrus.Debugf("removing %v from assigned names to image %q", nameToRemove, img.ID)
-			dest2, err := is.Transport.ParseStoreReference(b.store, "@"+imgID)
-			if err != nil {
-				return imgID, nil, "", fmt.Errorf("creating unnamed destination reference for image: %w", err)
-			}
-			dest = dest2
-		}
-		if options.IIDFile != "" {
-			if err = os.WriteFile(options.IIDFile, []byte("sha256:"+img.ID), 0644); err != nil {
-				return imgID, nil, "", err
+	dref := dest.DockerReference()
+	if dref != nil {
+		// Check if reference is on the storage
+		if _, err := b.store.Image(dref.String()); err == nil {
```

Comment:
@mtrmac Done, but question how does matching transport verifies that image is present on c/storage ?

Reply:
You’re right that just matching `transport.Name()` isn’t a presence check.

But this is `Commit`: the image was _just_ written to `dest` by the `retryCopyImage` line above, so the image must be present; if it isn’t, either something has gone horribly wrong. Or maybe the image was concurrently removed by another process in the meantime — but in that case we can’t correctly handle `IIDFile`, so this really should fail instead of silently continuing and ignoring that option.